Implement minimal EXIF and IPTC tag support

sebastiaan-lampo opened the pull request:
Updated version of PR #32.

Rebased on the updated version of HPS. I've added tests to cover the EXIF and IPTC additions.
Minor changes to the configuration tests are required as this PR introduces 3 potential new config file entries. These have been added to the config tests.

This feature allows:
- Back filling the copyright or caption data from an EXIF or IPTC metadata field if not defined in the album.yml
- Prepare for allowing access to all tags in the shortcode by making (a subset of) tags available as a property of the Photo object

sebastiaan-lampo committed:
Implement minimal EXIF and IPTC tag support

Use PIL's TAGS to deal with EXIF. Use
iptcinfo3 to deal with IPTC tags. Allows to
use these as fallback options for copyright
and caption of photos.

Will not change the default behaviour without
specifying the setting tag_map, exif or iptc to
the configuration file.

GjjvdBurg commented on hugophotoswipe/photo.py:
```diff
@@ -105,13 +108,105 @@ def _load_original_image(self):
         self._original_img = None
 
+    @property
+    def iptc(self):
```
What's the rationale for not having a `_load_iptc` method, similar to the `_load_exif` method?

sebastiaan-lampo replied:
Good point & implemented. 

The only downside is that it now always attempts to load IPTC. We always needed exif for orientation but we could lazy-load IPTC on first use. Right now it will simply load both.

The benefit of the code alignment may outweigh the performance penalty for users that don't use iptc data. I haven't tested the performance penalty yet.

GjjvdBurg replied:
I'm not sure I understand why you can't lazy load IPTC anyway. In `_get_tag_value`, you can load IPTC when needed.

Instead of:
```
o = getattr(self, obj.lower(), {})
```
you could use:
```
if obj.lower() == 'exif':
  return self.exif.get(t)
return self.iptc.get(t)
```
Then, in `def iptc(self)`, you can have:
```
if self._iptc is None:
  self._load_iptc()
return self._iptc
```

This means you won't have `_load_iptc` in the same place as `_load_exif`, but I think that's acceptable.

Let me know if you're interested in making this change. This has been open for a long time so if you're not, we could also just merge it and I can handle it instead.

GjjvdBurg commented on hugophotoswipe/photo.py:
```diff
@@ -105,13 +108,105 @@ def _load_original_image(self):
 
+    @property
+    def iptc(self):
+        if self._iptc is None:
```
You can do an early return here and save an indent:
```
if self._iptc is not None:
  return self._iptc
```
(small code style preference)

GjjvdBurg commented on docs/README.md:
```diff
@@ -42,6 +45,50 @@ unchanged. When a single number is given, the *maximum* dimension of the photo
+not be overridden, even after an update. Copyright is not yet used as this is currently 
+an album-level property.
+
+The format of the option is: `property: iptc/exif.tag`. Tag may include spaces. A full
```
Suggest to spell it out a bit more: "`<property_name>: iptc.<tag_name>` or `<property_name>:exif.<tag_name>` (fo example: `copyright: exif.ImageDescription`).". The same goes for the error message if the tag is improperly formatted (in `_get_tag_value`)

GjjvdBurg commented on hugophotoswipe/photo.py:
```diff
@@ -105,13 +108,105 @@ def _load_original_image(self):
+        assert tag is not None
+        try:
+            obj, t = tag.split('.')
+        except Exception as e:
```
I think the only error you can get here is a `ValueError`, right? Then you can just catch that instead of the general `Exception`.

sebastiaan-lampo replied:
Improved. Also added tests to check that malformed tag maps do indeed throw the expected error.

GjjvdBurg commented on hugophotoswipe/photo.py:
```diff
@@ -105,13 +108,105 @@ def _load_original_image(self):
+            return str(self._get_tag_value(settings.tag_map.get('copyright')))
+        return ""
+
+    def __getattribute__(self, attr):
```
I'm not a huge fan of overloading `__getattribute__` like this, especially because you're catching an `AttributeError` and then looking in the settings map. What if in the future I unintentionally introduce a class variable that's also an exif tag name? I'm not entirely sure where you're using this, but perhaps a `get_tag` method can work instead? Happy to discuss this if you'd like.

sebastiaan-lampo replied:
The reason for overloading `__getattribute__` is so that you can access those tags from within the hugo shortcode. That would allow you to expand the shortcode with, for example, `{{ .Get "aperture"}}`. I couldn't figure out another way to make that data accessible when processing the shortcode, except for explicitly defining each as a property. I'm definitely open to other ideas.

I do understand your objection though and as it isn't quite immediately related to the exif/iptc feature I'm removing this from this branch until we can find out a better solution.

GjjvdBurg replied:
Sorry for the slow response!

This makes sense, Here's one solution I can think of: require a prefix when used in a shortcode: `{{ .Get "exit:ApertureValue" }}` or `{{ .Get "iptc:ReleaseTime" }}`. Then you can overload the `__getattribute__` method by first checking for the `exif` or `iptc` prefixes, and retrieving them using `_get_tag_value`. If they don't start with either prefix, return `getattr(self, attr)`.

GjjvdBurg commented on hugophotoswipe/photo.py:
```diff
@@ -78,14 +82,13 @@ def original_image(self):
-        )
+        orientation = exif.get('Orientation')
         if orientation is None:
             print("Couldn't find orientation tag in ExifTags.TAGS")
```
This warning is triggered for me now when running the test cases, could this be because of your changes?

sebastiaan-lampo replied:
This was an old warning that sneaked back in. Removed.

GjjvdBurg reviewed (changes requested):
Thanks for adding this @sebastiaan-lampo! Very cool addition. I have some small comments from looking through the code and testing it, but hopefully these should be easy to fix. Thanks again and sorry for the delay! :)

sebastiaan-lampo committed:
Clarify exif|iptc tag format

sebastiaan-lampo committed:
Simplify behaviour when exif orientation is not defined

sebastiaan-lampo committed:
Load iptc data like exif at the same stage

sebastiaan-lampo committed:
Use pythonic list comprehension for _load_iptc

sebastiaan-lampo committed:
Improve error handling for improper tags

Also add test cases for improperly formatted tags.

sebastiaan-lampo committed:
Remove get_attribute overloader

Need a more robust solution to extract arbitrary attributes. For now, removing this from this branch.

sebastiaan-lampo committed:
Increase log level for iptcinfo to ERROR

iptcinfo spews many warnings about missing IPTC data, reading before image or decoding error. Most of the time this just seems to mean that the desired tags aren't present.

sebastiaan-lampo committed:
Fix bug in image rotation code

Note: Was not detected in test suite because all test images have exif['Orientation'] = None.

sebastiaan-lampo committed:
Fix bug in image rotation code

Note: Was not detected in test suite because all test images have exif['Orientation'] = None.

sebastiaan-lampo committed:
Merge remote-tracking branch 'sebastiaan-lampo/photo_exif' into photo_exif
